### ISSUE TITLE
Proposal: putting strings at the end of the document

### ISSUE THREAD

**Issue by michaeleisel** [msgpack/msgpack#265]
What if strings were placed at the end of the document? This would make it much faster to consume the document for a streaming parser, and the size of the document would be about the same. Although, many parsers have to consume the strings anyways, so this is just a thought.

**Comment by michaeleisel** [comment]
To elaborate, the string length would stay where it already was in the document, just the string itself would be moved

**Comment by tagomoris** [comment]
@michaeleisel I can't understand what you mean clearly.
Can you write your proposal using the syntax of [the spec file](https://github.com/msgpack/msgpack/blob/master/spec.md#str-format-family)?

**Comment by michaeleisel** [comment]
It's hard to write out completely using that syntax, but every str would move the `data` piece to the end of the document, right after the previous string's `data`. So e.g. for str 32 you'd go from:
```
+--------+--------+--------+--------+--------+========+
|  0xdb  |AAAAAAAA|AAAAAAAA|AAAAAAAA|AAAAAAAA|  data  |
+--------+--------+--------+--------+--------+========+
```
to
```
+--------+--------+--------+--------+--------+
|  0xdb  |AAAAAAAA|AAAAAAAA|AAAAAAAA|AAAAAAAA|
+--------+--------+--------+--------+--------+
```
and then `data` would go after the rest of the document, except for the `data` component of strings that come after it in the document. you'd also need a single number at the start of the document saying where the strings start

**Comment by markusschaber** [comment]
That makes it impossible to use it in a "real" streaming environment - where you want to parse the elements "on the fly", as they arrive. A writer would need to buffer and hold back all string contents until he has written the whole document (which might take a long time, effectively defeating the "streaming" behaviour), while a reader would also need to read the whole document before she could start to parse the strings.

**Comment by michaeleisel** [comment]
yes, that's true, it would depend on how the data is provided to the parser

**Comment by markusschaber** [comment]
What's your use case? What advantages do you see?

**Comment by michaeleisel** [comment]
i don't have a particular one at the moment for this, just thought i'd throw it out to brainstorm
